### ISSUE TITLE
Custom Emoji in Fediverse to Bluesky Display Names

### ISSUE THREAD

**Issue by hybridhavoc** [snarfed/bridgy-fed#1104]
In Mastodon and other Fediverse applications it is possible to include a custom emoji from one's server in the display name. While those applications support this use, Bluesky of course does not. This results in the shortcode for the custom emoji coming over in the display name.

I don't know if this is the best way to handle this exactly, but I wonder if it might be possible, upon Bluesky profile creation / update to strip custom emoji shortcodes from the display name.

**Comment by snarfed** [comment]
Yes! Agreed, these are awkward. I don't love the idea of removing them, but I don't have a better idea, so we probably should.

**Comment by snarfed** [comment]
Out of curiosity, what do you think we should do about custom emoji shortcodes in post text? Strip them there too?

**Comment by qazmlp** [comment]
> Out of curiosity, what do you think we should do about custom emoji shortcodes in post text? Strip them there too?

No, definitely keep them, or at most make it a setting if you introduce that at some point.

Even if they don't transfer 'properly' and it looks a bit ugly, having their text representation is still better than removing them silently. I think readers will also generally understand what's happenening because emoji shortcode use is somewhat generally common now thanks to being very discoverable in Discord.

**Comment by hybridhavoc** [comment]
> Out of curiosity, what do you think we should do about custom emoji shortcodes in post text? Strip them there too?

For some reason I'm less inclined to modify the contents of a post. Though I will say that if a post needs to be abbreviated for being too long for Bluesky, it might be preferable to not break in the middle of a short code.

**Comment by snarfed** [comment]
Interesting! I have to admit, I don't quite get the intuition yet for removing these shortcodes from display names but keeping them in posts. Is it that we think they're more purely decorative in display names, but may have semantic meaning more often in posts?

**Comment by hybridhavoc** [comment]
> Is it that we think they're more purely decorative in display names, but may have semantic meaning more often in posts?

Yes, at least just speaking for myself. I like to sometimes add custom emoji to my display name for decoration in a way that isn't a big deal and isn't intended to communicate much of significance. Whereas an emoji in the body of a post is more likely being used for signifying tone or replacing a word.

It may also be a matter of impact. My display name is seen above every post I make, so having what may come across as gibberish there all the time is more impactful than if it appears in an occasional post.

**Comment by ygg2** [comment]
^ Definitely please don't remove the emoji codes! Many users especially on Misskey based software replace words with emojis, so something like https://misskey.io/notes/9te0y2qi2ie7098x (I got locked out of X because of Firefox) would not make sense, and posts like https://misskey.io/notes/9u5m67deo9tk01ee (Goodnight Misskey! See you) would be blank.

**Comment by snarfed** [comment]
Thanks all!

Next step is implementation. These aren't standardized afaik, and I doubt that all fediverse software with shortcodes uses the same HTML for them (if any). I don't know if I can come up with a regex to find and remove them, even if only from display names, that won't occasionally have false positives and remove other text too. 😕

**Comment by Tamschi** [comment]
Mastodon declares these as `"type": "Emoji", "name": ":…:"`` objects in the `"tag": [ … ]` array of the Actor, same as on statuses/Notes if I'm not mistaken.

I think you can just search&remove the custom emojis' `name`s from the Actor's `name`, then trim to remove left-over spaces.

**Comment by snarfed** [comment]
Another idea, from @letydoesstuff in #1553. I like it!

> ActivityPub supports custom emoji while ATProto does not. Given that and character limits on Bluesky, I feel :custom_emoji_shortcodes: would be better served as � tofu replacement characters instead.
>
> CW: Marked Sensitive Media Below
>
> But that [this post](https://doesstuff.social/@lety/113541183145762153) currently bridges over like [this silliness](https://bsky.app/profile/lety.doesstuff.social.ap.brid.gy/post/3lbqhclxres32) seems less than ideal.

**Comment by Tamschi** [comment]
The [current Bluemoji draft spec](https://github.com/aendra-rininsland/bluemoji/pull/1/files?short_path=b76a8c4#diff-b76a8c46c7b5cc6d5a145b94544754096022de4036b13ff259b78033edd2d7c5) recommends `◌` as fallback character.

(Once that stabilises, it's most likely worth looking at in terms of bridging custom emojis.)

**Comment by letydoesstuff** [comment]
I admire the idea of forming consensus among OSS projects, but Bluemoji has not been worked on in seven months while snarfed actively develops Bridgy Fed, which sees increasing use.

I would like to raise my proposal at snarfed/bridgy-fed#1553 once again and suggest that custom emoji be replaced with � tofu replacement characters.

Keeping :custom-emoji-in-shortcode-form: eats up valuable AT Protocol character count and makes emoji-only posts unnecessarily verbose, removing custom emoji altogether strips all evidence of now missing characters and creates further issues with emoji-only posts and responses, and the median internet user does not know what character encodings are in general.

The Unicode standard is to use � in replacement of unrenderable characters, and that is what custom emoji from ActivityPub are when bridged onto the AT Protocol.

**Comment by snarfed** [comment]
I have a first pass at this in https://github.com/snarfed/granary/pull/1047, but I don't think it's ready to merge yet. I think it also needs to handle other tags with indices and shuffle them over too.
